lib/ruby_lsp/rbs_document.rb

```diff
@@ -0,0 +1,41 @@
+# typed: strict
+# frozen_string_literal: true
+
+module RubyLsp
+  class RBSDocument < Document
+    extend T::Sig
+    extend T::Generic
+
+    ParseResultType = type_member { { fixed: T::Array[RBS::AST::Declarations::Base] } }
+
+    sig { params(source: String, version: Integer, uri: URI::Generic, encoding: Encoding).void }
+    def initialize(source:, version:, uri:, encoding: Encoding::UTF_8)
+      @syntax_error = T.let(false, T::Boolean)
+      super
+    end
+
+    sig { override.returns(ParseResultType) }
+    def parse
+      return @parse_result unless @needs_parsing
+
+      @needs_parsing = false
```

Comment:
Indeed, it's counterintuitive. This is a performance optimization. We only try to parse documents if a new edit has been made.

The reason is because if you have a parse error, until the user makes a new edit, there's no possibility of that error going away. So trying to re-parse the document only needs to happen if a new edit was made - not if parsing succeeded or not.